app-runner/Private/DeviceProviders/AdbProvider.ps1

```diff
@@ -0,0 +1,521 @@
+        return @{
+            Platform       = $this.Platform
+            ExecutablePath = $ExecutablePath
+            Arguments      = $Arguments
+            StartedAt      = $startTime
+            FinishedAt     = Get-Date
+            Output         = $formattedLogs
+            ExitCode       = $exitCode
+        }
+    }
+
+    # Helper method to wait for process with retry
+    [string] WaitForProcess([string]$packageName, [int]$timeoutSeconds) {
+        Write-Debug "$($this.Platform): Waiting for process: $packageName"
+
+        for ($i = 0; $i -lt $timeoutSeconds; $i++) {
+            try {
+                $PSNativeCommandUseErrorActionPreference = $false
+                $pidOutput = $this.InvokeCommand('pidof', @($this.DeviceSerial, $packageName))
+                $PSNativeCommandUseErrorActionPreference = $true
+
+                if ($pidOutput) {
+                    $processId = $pidOutput.ToString().Trim()
+                    if ($processId -match '^\d+$') {
+                        return $processId
+                    }
+                    else {
+                        Write-Warning "Unexpected pidof output: $processId"
+                    }
+                }
+            }
+            catch {
+                # Process not found yet, continue waiting
+                Write-Debug "$($this.Platform): Process not found yet (attempt $i/$timeoutSeconds)"
+            }
+
+            Start-Sleep -Seconds 1
+        }
+
+        return $null
+    }
+
+    [hashtable] GetDeviceLogs([string]$LogType, [int]$MaxEntries) {
+        Write-Debug "$($this.Platform): Getting device logs (type: $LogType, max: $MaxEntries)"
+
+        $logs = $this.InvokeCommand('logcat', @($this.DeviceSerial))
+        $formattedLogs = Format-LogcatOutput -LogcatOutput $logs
+
+        if ($MaxEntries -gt 0) {
+            $formattedLogs = $formattedLogs | Select-Object -Last $MaxEntries
+        }
+
+        return @{
+            Platform  = $this.Platform
+            LogType   = $LogType
+            Logs      = $formattedLogs
+            Count     = $formattedLogs.Count
+            Timestamp = Get-Date
+        }
+    }
+
+    [void] TakeScreenshot([string]$OutputPath) {
+        Write-Debug "$($this.Platform): Taking screenshot to: $OutputPath"
+
+        # Ensure output directory exists
+        $directory = Split-Path $OutputPath -Parent
+        if ($directory -and -not (Test-Path $directory)) {
+            New-Item -Path $directory -ItemType Directory -Force | Out-Null
+        }
+
+        # Use intermediate file on device to avoid binary data corruption in stdout capture
+        $tempDevicePath = "/sdcard/temp_screenshot_$([Guid]::NewGuid()).png"
+
+        try {
+            # Capture to temp file on device
+            $this.InvokeCommand('screenshot-file', @($this.DeviceSerial, $tempDevicePath))
+
+            # Pull file from device
+            $this.InvokeCommand('pull', @($this.DeviceSerial, $tempDevicePath, $OutputPath))
```

Comment:
This should use CopyDeviceItem to prevent code duplication:
- the command call (this line) - not really a problem though
- creating the output directory